src/lib/protocols/knxnet_ip.c

```diff
@@ -0,0 +1,107 @@
+ * knxnet_ip.c
+ *
+ * KNXnet/IP
+ * 
+ * Copyright (C) 2024 - ntop.org
+ * Copyright (C) 2024 - V.G <jacendi@protonmail.com>
+ *
+ * This file is part of nDPI, an open source deep packet inspection
+ * library based on the OpenDPI and PACE technology by ipoque GmbH
+ *
+ * nDPI is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * nDPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with nDPI.  If not, see <http://www.gnu.org/licenses/>.
+ * 
+ */
+
+#include "ndpi_protocol_ids.h"
+
+#define NDPI_CURRENT_PROTO NDPI_PROTOCOL_KNXNET_IP
+
+#include "ndpi_api.h"
+#include "ndpi_private.h"
+
+enum knx_service_families {
+  KNX_SERVICE_CORE                    = 0x02,
+  KNX_SERVICE_MANAGEMENT              = 0x03,
+  KNX_SERVICE_TUNNELING               = 0x04,
+  KNX_SERVICE_ROUTING                 = 0x05,
+  KNX_SERVICE_REMOTE_LOGGING          = 0x06,
+  KNX_SERVICE_REMOTE_DIAG_AND_CONFIG  = 0x07,
+  KNX_SERVICE_OBJECT_SERVER           = 0x08,
+  KNX_SERVICE_SECURITY                = 0x09
+};
+
+static inline int is_valid_knxnet_ip_service_id(u_int16_t service_id) {
+  u_int8_t service_family = service_id >> 8;
+  u_int8_t service_code = service_id & 0xFF;
+
+  switch (service_family)
+  {
+    case KNX_SERVICE_CORE:
+      return (service_code > 0 && service_code < 0xD);
+    case KNX_SERVICE_MANAGEMENT:
+      return (service_code >= 0x10 && service_code <= 0x11);
+    case KNX_SERVICE_TUNNELING:
+      return (service_code >= 0x20 && service_code <= 0x25);
+    case KNX_SERVICE_ROUTING:
+      return (service_code >= 0x30 && service_code <= 0x33);
+    case KNX_SERVICE_REMOTE_DIAG_AND_CONFIG:
+      return (service_code >= 0x40 && service_code <= 0x43);
+    case KNX_SERVICE_SECURITY:
+      return (service_code >= 0x50 && service_code <= 0x55);
+    default:
+      return 0;
+  }
+}
+
+static void ndpi_search_knxnet_ip(struct ndpi_detection_module_struct *ndpi_struct,
+                                  struct ndpi_flow_struct *flow)
+{
+  struct ndpi_packet_struct const * const packet = &ndpi_struct->packet;
+
+  NDPI_LOG_DBG(ndpi_struct, "search KNXnet/IP\n");
+
+  if ((packet->payload_packet_len < 10) || (packet->payload[0] != 0x06) ||
+      (packet->payload[1] != 0x10))
+  {
+    NDPI_EXCLUDE_PROTO(ndpi_struct, flow);
+    return;
+  }
+
+  u_int16_t service_id = ntohs(get_u_int16_t(packet->payload, 2));
+  u_int16_t total_length = ntohs(get_u_int16_t(packet->payload, 4));
+
+  if (is_valid_knxnet_ip_service_id(service_id) && total_length == packet->payload_packet_len)
```

Comment:
Could we do something like, in pseudo-code:
```
if udp --> total_length == packet->payload_packet_len
else --> total_length <= packet->payload_packet_len
```
to try to handle also the TCP sample? If so, could you update the trace with the TCP flow, please?

Reply:
> Could we do something like, in pseudo-code:
> 
> ```
> if udp --> total_length == packet->payload_packet_len
> else --> total_length <= packet->payload_packet_len
> ```
> 
> to try to handle also the TCP sample? If so, could you update the trace with the TCP flow, please?

Yeah, sure. I read the protocol specs and found that KNXnet/IP has the feature to pack multiple messages in a single TCP packet.

In this case we can use the `Total Length` field of the first message as an offset to check if there is another KNXnet/IP header (i.e. bytes `06 10` and another `Service ID`).